llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
   return success();
 }
 
+LogicalResult
+TopkOp::reifyResultShapes(OpBuilder &b,
+                          ReifiedRankedShapedTypeDims &reifiedReturnShapes) {
+  return cast<LinalgExtOp>(getOperation())
+      .reifyResultShapes(b, reifiedReturnShapes);
+}
+
+//===----------------------------------------------------------------------===//
+// PackOp
+//===----------------------------------------------------------------------===//
+
+// Return true if each element in `dimsPos` is >= 0 and < rank.
+static bool isInBound(ArrayRef<int64_t> dimsPos, int64_t rank) {
+  return llvm::all_of(
+      dimsPos, [rank](int64_t dimPos) { return dimPos >= 0 && dimPos < rank; });
+}
+
+// Interchange `elements` starting at offset `offset` based on the indexes in
+// `interchangeVector`.
+template <typename T>
+static SmallVector<T> interchange(ArrayRef<T> elements,
+                                  ArrayRef<int64_t> interchangeVector,
+                                  int64_t offset) {
+  SmallVector<T> rearrangedElements = llvm::to_vector(elements);
+  if (interchangeVector.empty())
+    return rearrangedElements;
+  assert((rearrangedElements.size() - offset) == interchangeVector.size() &&
+         "number of elements must equal number of permutations");
+  for (int64_t idx = 0, end = interchangeVector.size(); idx < end; idx++)
+    rearrangedElements[interchangeVector[idx] + offset] =
+        elements[idx + offset];
+  return rearrangedElements;
+}
+
+// Infer result/output type given the input and the tile sizes.
+ShapedType PackOp::inferResultType() {
+  DenseMap<int64_t, int64_t> tileAndPosMapping = getDimAndStaticTileMapping();
+  SmallVector<int64_t> inferredShape;
+  inferredShape.reserve(getOutputRank());
+  ShapedType inputType = getInputType();
+  int64_t rank = getInputRank();
+
+  // tile loop.
+  for (auto i : llvm::seq<int64_t>(0, rank)) {
+    if (tileAndPosMapping.count(i)) {
+      if (inputType.isDynamicDim(i) ||
+          tileAndPosMapping[i] == ShapedType::kDynamicSize)
```

Comment:
If you used `getDimAndTileMapping` you could change this

```
for (auto i : ) {
   if (tileAndPosMapping.count(i)) {
      Optional<int64_t> tileSize = getConstantIntValue(tileAndPosMapping.lookup(i));
      // Drop all `== ShapedType::kDynamicSize` everywhere.
   }
}
```

The `getConstantIntValue` is defined in `StaticValueUtils.h` in core